python/cuml/tests/test_umap_hypothesis.py

```diff
@@ -0,0 +1,616 @@
+#
+# SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION.
+# SPDX-License-Identifier: Apache-2.0
+#
+
+"""
+Hypothesis-based property testing for UMAP simplicial set embedding.
+
+This test suite uses property-based testing to thoroughly validate the
+simplicial set embedding implementation against the reference UMAP
+implementation across a wide range of parameters and dataset configurations.
+"""
+
+import os
+import sys
+
+import cupy as cp
+import hypothesis
+import numpy as np
+import pytest
+from hypothesis import example, given, settings
+from hypothesis import strategies as st
+from scipy.sparse import csr_matrix
+
+# Reference UMAP implementation
+from umap.umap_ import find_ab_params
+from umap.umap_ import simplicial_set_embedding as ref_simplicial_set_embedding
+
+import cuml.datasets
+
+# cuML implementation
+from cuml.manifold.simpl_set import (
+    fuzzy_simplicial_set as cu_fuzzy_simplicial_set,
+)
+from cuml.manifold.simpl_set import (
+    simplicial_set_embedding as cu_simplicial_set_embedding,
+)
+
+# Check if cuvs is available (required for this test)
+try:
+    import cuvs  # noqa: F401
+
+    HAS_CUVS = True
+except ImportError:
+    HAS_CUVS = False
+
+# Only import umap_metrics if cuvs is available (umap_metrics depends on cuvs)
+if HAS_CUVS:
+    umap_dev_tools_path = os.path.join(
+        os.path.dirname(__file__), "..", "umap_dev_tools"
+    )
+    if os.path.exists(umap_dev_tools_path):
+        sys.path.insert(0, umap_dev_tools_path)
+
+    try:
+        from umap_metrics import (  # noqa: E402
+            _build_knn_with_cuvs,
+            compute_simplicial_set_embedding_metrics,
+            procrustes_rmse,
+        )
+    except ImportError as e:
+        raise ImportError(
+            f"Failed to import umap_metrics. Ensure the umap_dev_tools module is "
+            f"available at {umap_dev_tools_path}. Original error: {e}"
+        ) from e
+else:
+    # Define dummy functions to avoid NameError, tests will be skipped anyway
+    _build_knn_with_cuvs = None
+    compute_simplicial_set_embedding_metrics = None
+    procrustes_rmse = None
+
+
+# Custom dataset wrapper with concise repr to avoid printing giant arrays
+class DatasetWrapper:
+    """Wrapper for dataset to provide concise string representation."""
```

Comment:
It seems like this wrapper is doing more than that.

Reply:
I improved the description/documentation.